mitodl/ocw-studio
# add ckeditor media embed plugin

#### Pre-Flight checklist


#### What are the relevant tickets?

for #39 

#### What's this PR do?

This adds the [MediaEmbed](https://ckeditor.com/docs/ckeditor5/latest/features/media-embed.html) plugin to our CKEditor setup. In order to support this we also need to extend our Markdown layer, so I added a new framework for doing that involving a new type of plugin.

It is configured right now to support only YouTube embeds - if we want to support more providers we'll need to extend the Markdown translation part to support other shortcodes.

#### How should this be manually tested?

You should be able to add a YouTube video to the editor. It should show up in the output as a Hugo-compatible shortcode. If you start the editor with a youtube shortcode in the input data it should up in the editor as a preview without any issues.

Also! Definitely review the architecture carefully around customizing the markdown layer. It's set up so each custom plugin will now also define a plugin that extends `MarkdownSyntaxPlugin`, and this plugin basically just defines the showdown and turndown rules that we want to use for the plugin. This will let us keep the code for the markdown stuff right with the rest of the code for the plugin, and will also ensure that markdown support is added to the editor at the same time as the rest of the new plugin.

#### Screenshots (if appropriate)

![Screenshot from 2021-03-04 15-30-19](https://user-images.githubusercontent.com/6207644/110027420-f1ca3e00-7cff-11eb-96f7-a66e6cc685af.png)
![Screenshot from 2021-03-04 15-30-25](https://user-images.githubusercontent.com/6207644/110027426-f42c9800-7cff-11eb-9fa0-cde17504d90b.png)
